### PR TITLE
Tidy propagateDashboardParameters

### PR DESCRIPTION
**Pull request by calherries**
While fixing #25521 I found this particular part of the code quite complex to run through due to all the conditionals.

I've refactored it a bit for readability.

### CODE REVIEW

File: frontend/src/metabase/query_builder/actions/core/initializeQB.ts [hunk 1]
```diff
@@ -247,17 +247,17 @@ async function handleQBInit(
     return;
   }
 
-  if (hasCard) {
-    await handleDashboardParameters(card, {
+  if (deserializedCard?.dashcardId) {
+    card = await propagateDashboardParameters({
+      card,
       deserializedCard,
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/initializeQB.ts:253]: We were checking this inside the `handleDashboardParameters` before, but IMO for clarity, the function should only be called if deserializedCard is a dashcard. Otherwise, there's clearly no need to propagate dashboard parameters.
https://github.com/metabase/metabase/pull/25539/files#diff-74a80e24cd9cea4e9a716c02394f89df3331c13a1227fa4b9d7a76d81165928aL131

File: frontend/src/metabase/query_builder/actions/core/initializeQB.ts [hunk 2]
```diff
@@ -247,17 +247,17 @@ async function handleQBInit(
...
     });
-  } else {
-    if (options.metric) {
-      uiControls.isShowingSummarySidebar = true;
-    }
+  }
+
+  if (!hasCard && options.metric) {
+    uiControls.isShowingSummarySidebar = true;
   }
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/initializeQB.ts:261]: Shuffled the `hasCard` check down here

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 3]
```diff
@@ -29,33 +29,25 @@ type QueryParams = BlankQueryOptions & {
...
 
-function checkShouldPropagateDashboardParameters({
+function shouldPropagateDashboardParameters({
   cardId,
   deserializedCard,
   originalCard,
 }: {
   cardId?: number;
-  deserializedCard?: Card;
+  deserializedCard: Card;
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:38]: Now we have a stronger precondition: `deserializedCard` is definitely non-null

File: frontend/src/metabase/query_builder/actions/core/initializeQB.ts [hunk 4]
```diff
@@ -247,17 +247,17 @@ async function handleQBInit(
...
-      getState,
     });
-  } else {
-    if (options.metric) {
-      uiControls.isShowingSummarySidebar = true;
-    }
+  }
+
+  if (!hasCard && options.metric) {
+    uiControls.isShowingSummarySidebar = true;
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/initializeQB.ts:260]: I shuffled the `hasCard` check down to here. I believe these are orthogonal concerns anyway.

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 5]
```diff
@@ -107,34 +97,37 @@ export function getParameterValuesForQuestion({
...
-    originalCard,
-  });
-  if (shouldPropagateParameters && deserializedCard?.dashcardId) {
+  if (
+    shouldPropagateDashboardParameters({
+      cardId,
+      deserializedCard,
+      originalCard,
+    })
+  ) {
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:126]: simple substitution

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 6]
```diff
@@ -144,4 +137,5 @@ export async function handleDashboardParameters(
     card.dashboardId = dashboardId;
     card.dashcardId = dashcardId;
   }
+  return card;
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:140]: for clarity, since we are mutating card here

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 7]
```diff
@@ -29,33 +29,27 @@ type QueryParams = BlankQueryOptions & {
...
+    // Either the card is saved and deserializedCard has parameters
+    (cardId && deserializedCard.parameters) ||
+      // Or deserializedCard is equivalent to originalCard, except for its parameters
+      (originalCard &&
+        cardIsEquivalent(deserializedCard, originalCard, {
+          checkParameters: false,
+        }) &&
+        !cardIsEquivalent(deserializedCard, originalCard, {
+          checkParameters: true,
+        })),
```
Comment (calherries) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:51]: This is exactly the same as before, except compressed into a boolean expression.

It's just missing the first check:
```
if (!deserializedCard) {
    return false;
  }
```

Comment (gusaiani) [reply frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:51]: This block I find more friendly in the original structure.

Particularly `return equalCards && differentParameters`, that I find highly declarative.

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 8]
```diff
@@ -107,41 +99,44 @@ export function getParameterValuesForQuestion({
...
-  },
-) {
+/**
+ * Merges .parameters, .dashboardId, and .dashcardId props from deserializedCard into card.
+ * Sets an error page if there have been permissions or data changes to a dashboard such that:
+ *  - If the user loses permissions to view the dashboard, the user will be navigated to an unauthed screen.
+ *  - If the card is removed from the dashboard or some of the parameters mapped to it have been removed,
+ *    the user will be navigated to an unauthed screen.
+ * See https://github.com/metabase/metabase/pull/19300 for the origin of the error handling.
+ */
```
Comment (gusaiani) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:109]: Much appreciated.

**Review by gusaiani — APPROVED** [review]
Looks good to me overall.

One comment inlined, please take a look.

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts (hunk 7 above)
Comment (kulyk) [reply frontend/src/metabase/query_builder/actions/core/parameterUtils.ts]: +1 to Gus, we should avoid complex nested conditions like that one; I'd vote for breaking it down. Refactoring these actions for readability and untangling deep nesting came with a cost; we shouldn't get back to it again

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts (hunk 8 above)
Comment (kulyk) [reply frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:114]: +1, thanks @calherries 🙏

**Review by kulyk — CHANGES_REQUESTED** [review]
Overall looks good, appreciate the clean-up!

@calherries could you please untangle a complex conditional me and Gus have highlighted? Also, if that makes sense to do here, I'd really appreciate if we can extract [this block](https://github.com/metabase/metabase/blob/c4727325b59190cde0f814705fef8406194f60ca/frontend/src/metabase/query_builder/actions/core/initializeQB.ts#L295) out of `handleInitQB`. This action used to be insanely nested and hard to follow before. I'm worried adding a few nested blocks like that over the next few months may bring us to the point we want to refactor it once again.

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 9]
```diff
@@ -29,33 +32,29 @@ type QueryParams = BlankQueryOptions & {
...
   cardId?: number;
-  deserializedCard?: Card;
+  deserializedCard: Card;
   originalCard?: Card;
-}) {
-  if (!deserializedCard) {
-    return false;
-  }
+}): boolean {
+  // Either the card is saved and deserializedCard has parameters
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:44]: I _think_ this comment repeats what the code itself can communicate, i.e. seems to be more like a [_how_ comment rather than _a why_ comment](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/)

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts [hunk 10]
```diff
@@ -29,33 +32,29 @@ type QueryParams = BlankQueryOptions & {
...
   if (cardId && deserializedCard.parameters) {
     return true;
-  }
-  if (!originalCard) {
+  } else if (!originalCard) {
     return false;
+  } else {
+    // Or deserializedCard is equivalent to originalCard, except for its parameters
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:50]: Same thoughts as about the comment above

**Comment by kulyk** [comment]
@daltojohnso I'd appreciate if you could take a look at the parameter-related changes here

File: frontend/src/metabase/query_builder/actions/core/parameterUtils.ts (hunk 9 above)
Comment (calherries) [reply frontend/src/metabase/query_builder/actions/core/parameterUtils.ts:44]: Good point, I'll delete these

**Comment by calherries** [comment]
@kulyk 

> if that makes sense to do here, I'd really appreciate if we can extract [this block](https://github.com/metabase/metabase/blob/c4727325b59190cde0f814705fef8406194f60ca/frontend/src/metabase/query_builder/actions/core/initializeQB.ts#L295) out of handleInitQB.

Yes, I think that's possible in another PR. I'll do it now.

**Comment by calherries** [comment]
@kulyk I've followed up on the above in this PR https://github.com/metabase/metabase/pull/25576

**Review by daltojohnso — APPROVED** [review]

**Review by kulyk — APPROVED** [review]
LGTM, sorry for a bit of a lag from my side
